src/kdf.rs

```diff
@@ -0,0 +1,351 @@
+                return err_rv!(CKR_MECHANISM_INVALID);
+            }
+            _ => return err_rv!(CKR_MECHANISM_INVALID),
+        };
+        Ok(Operation::Derive(Box::new(kdf)))
+    }
+}
+
+pub fn register(mechs: &mut Mechanisms, _: &mut ObjectFactories) {
+    Sp800Operation::register_mechanisms(mechs);
+}
+
+#[derive(Debug)]
+struct Sp800CounterFormat {
+    defined: bool,
+    le: bool,
+    bits: usize,
+}
+
+#[derive(Debug)]
+struct Sp800DKMLengthFormat {
+    method: CK_ULONG,
+    le: bool,
+    bits: usize,
+}
+
+#[derive(Debug)]
+enum Sp800Params {
+    Iteration(Sp800CounterFormat),
+    Counter(Sp800CounterFormat),
+    ByteArray(Vec<u8>),
+    DKMLength(Sp800DKMLengthFormat),
+}
+
+#[derive(Debug)]
+struct Sp800Operation {
+    mech: CK_MECHANISM_TYPE,
+    prf: CK_MECHANISM_TYPE,
+    finalized: bool,
+    in_use: bool,
+    params: Vec<Sp800Params>,
+    iv: Vec<u8>,
+    addl_drv_keys: Vec<CK_DERIVED_KEY>,
+    addl_objects: Vec<Object>,
+}
+
+unsafe impl Send for Sp800Operation {}
+unsafe impl Sync for Sp800Operation {}
+
+impl Sp800Operation {
+    fn register_mechanisms(mechs: &mut Mechanisms) {
+        mechs.add_mechanism(
+            CKM_SP800_108_COUNTER_KDF,
+            Box::new(Sp800KDFMechanism {
+                info: CK_MECHANISM_INFO {
+                    ulMinKeySize: 0,
+                    ulMaxKeySize: std::u32::MAX as CK_ULONG,
+                    flags: CKF_DERIVE,
+                },
+                mech_type: CKM_SP800_108_COUNTER_KDF,
+            }),
+        );
+        mechs.add_mechanism(
+            CKM_SP800_108_FEEDBACK_KDF,
+            Box::new(Sp800KDFMechanism {
+                info: CK_MECHANISM_INFO {
+                    ulMinKeySize: 0,
+                    ulMaxKeySize: std::u32::MAX as CK_ULONG,
+                    flags: CKF_DERIVE,
+                },
+                mech_type: CKM_SP800_108_FEEDBACK_KDF,
+            }),
+        );
+    }
+
+    fn parse_counter_format(
+        p: &CK_PRF_DATA_PARAM,
+    ) -> KResult<Sp800CounterFormat> {
+        if p.ulValueLen == 0 && p.pValue == std::ptr::null_mut() {
+            return Ok(Sp800CounterFormat {
+                defined: false,
+                le: false,
+                bits: 16,
+            });
+        }
+        if p.ulValueLen as usize
+            != ::std::mem::size_of::<CK_SP800_108_COUNTER_FORMAT>()
+        {
+            return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+        }
+        let cf = unsafe { *(p.pValue as *const CK_SP800_108_COUNTER_FORMAT) };
+        Ok(Sp800CounterFormat {
+            defined: true,
+            le: match cf.bLittleEndian {
+                0 => false,
+                1 => true,
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+            },
+            bits: match cf.ulWidthInBits {
+                8 | 16 | 24 | 32 => cf.ulWidthInBits as usize,
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
```

Comment:
This restricts the max size of the format to 32 bits, which is probably fine, but we should note that. So every counter has to be 32 bits or less.

Reply:
This is directly from the sp800-108 spec:
```
10) r – An integer, smaller or equal to 32, that indicates the length of the binary
representation of the counter i.

```and NSS does the same.